### ISSUE TITLE
Unable to connect to 4.0 instance with 4.0 module due to thrown certification exception

### ISSUE THREAD

**Issue by danlemire** [HewlettPackard/POSH-HPOneView#303]
### Expected Behavior
Established connection with a OneView Composer

### Actual Behavior
Exception thrown for certificate hostname not matching subject or SAN, connection cannot be established.
"VERBOSE: [SEND-HPOVREQUEST] InnerException Message: The hostname used to connect does not match the Subject or SAN of the provided host certificate.  Throw exception."

### Steps to reproduce
After installing HPOneView400, attempt to connect to a OneView 4 Synergy Composer (configured with a self-signed certification (complete with both an FQDN and alternate names that include the shortname and ip address)
Connect-HPOVMgmt -Hostname name.domain.suffix -UserName Administrator -Password blahblahblah1 -verbose

```PowerShell
# 1. Capture verbose output using the Get-HPOVCommandTrace for HPOneView.310 or newer, or append the -Verbose switch to your Cmdlet call.

VERBOSE: [CONNECT-HPOVMGMT] Bound PS Parameters: 
Name                           Value                                                                                                                                                                                                                 
----                           -----                                                                                                                                                                                                                 
UserName                       Administrator                                                                                                                                                                                                         
Verbose                        True                                                                                                                                                                                                                  
Hostname                       hostname.domain.suffix                                                                                                                                                                                          
Password                       [*****REDACTED******]                                                                                                                                                                                                 



VERBOSE: [CONNECT-HPOVMGMT] Creating Session Container
VERBOSE: [CONNECT-HPOVMGMT] Checking X-API Version
VERBOSE: [SEND-HPOVREQUEST] BEGIN
VERBOSE: [SEND-HPOVREQUEST] Bound PS Parameters: 
Key      Value                       
---      -----                       
Hostname hostname.domain.suffix
uri      /rest/version               



VERBOSE: [SEND-HPOVREQUEST] Process
VERBOSE: [SEND-HPOVREQUEST] Hostname value: hostname.domain.suffix

VERBOSE: [SEND-HPOVREQUEST] Filtering for Connection Object via String: hostname.domain.suffix
VERBOSE: [SEND-HPOVREQUEST] Processing 'hostname.domain.suffix' appliance connection request. 1 of 1
VERBOSE: [SEND-HPOVREQUEST] Requested URI '/rest/version' to 'hostname.domain.suffix'
VERBOSE: [SEND-HPOVREQUEST] We have reached the URI Whitelist condition block. Unauth request allowed for '/rest/version'.
VERBOSE: [RESTCLIENT] Building new [System.Net.HttpWebRequest] object for GET https://hostname.domain.suffix/rest/version
VERBOSE: [SEND-HPOVREQUEST] Request: GET https://hostname.domain.suffix/rest/version
VERBOSE: [SEND-HPOVREQUEST] Request Header 1: User-Agent = HPEOneview_PowerShell/4.00
VERBOSE: [SEND-HPOVREQUEST] Request Header 2: Content-Type = application/json
VERBOSE: [SEND-HPOVREQUEST] Request Header 3: Accept = application/json
VERBOSE: [SEND-HPOVREQUEST] Request Header 4: X-API-Version = 600
VERBOSE: [SEND-HPOVREQUEST] Request Header 5: accept-language = en_US
VERBOSE: [SEND-HPOVREQUEST] Request Header 6: accept-encoding = gzip, deflate
VERBOSE: [SEND-HPOVREQUEST] Request Header 7: auth = [*****REDACTED******]
VERBOSE: [SEND-HPOVREQUEST] Net.WebException Error caught
VERBOSE: [SEND-HPOVREQUEST] Exception Object: 

PSMessageDetails      : 
Exception             : System.Net.WebException: The underlying connection was closed: An unexpected error occurred on a receive. ---> HPOneView.Library.ApplianceTransportException: The hostname used to connect does not match the Subject or SAN 
                        of the provided host certificate.  Throw exception.
                           at HPOneView.PKI.SslValidation.CertificateValidationCallBack(Object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
                           at System.Net.ServerCertValidationCallback.Callback(Object state)
                           at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
                           at System.Net.ServerCertValidationCallback.Invoke(Object request, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
                           at System.Net.Security.SecureChannel.VerifyRemoteCertificate(RemoteCertValidationCallback remoteCertValidationCallback, ProtocolToken& alertToken)
                           at System.Net.Security.SslState.CompleteHandshake(ProtocolToken& alertToken)
                           at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)
                           at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
                           at System.Net.TlsStream.ProcessAuthentication(LazyAsyncResult result)
                           at System.Net.TlsStream.Write(Byte[] buffer, Int32 offset, Int32 size)
                           at System.Net.ConnectStream.WriteHeaders(Boolean async)
                           --- End of inner exception stack trace ---
                           at System.Net.HttpWebRequest.GetResponse()
                           at CallSite.Target(Closure , CallSite , WebRequest )
TargetObject          : 
CategoryInfo          : NotSpecified: (:) [], WebException
FullyQualifiedErrorId : WebException
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Send-HPOVRequest<Process>, C:\Program Files\WindowsPowerShell\Modules\HPOneView.400\4.0.1554.2229\HPOneView.400.psm1: line 4702
                        at Connect-HPOVMgmt<Begin>, C:\Program Files\WindowsPowerShell\Modules\HPOneView.400\4.0.1554.2229\HPOneView.400.psm1: line 6459
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}




VERBOSE: [SEND-HPOVREQUEST] Exception Message: The underlying connection was closed: An unexpected error occurred on a receive.
VERBOSE: [SEND-HPOVREQUEST] InnerException FullyQualifiedErrorId: ApplianceTransportException
VERBOSE: [SEND-HPOVREQUEST] InnerException Message: The hostname used to connect does not match the Subject or SAN of the provided host certificate.  Throw exception.
VERBOSE: [NEW-ERRORRECORD] Building ErrorRecord object
VERBOSE: [SEND-HPOVREQUEST] Cleaning up HttpWebRequest
Connect-HPOVMgmt : The hostname used to connect does not match the Subject or SAN of the provided host certificate.  Throw exception.
At line:1 char:1
+ Connect-HPOVMgmt -Hostname hostname.domain.suffix -UserName Adm ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (Hostname:String) [Connect-HPOVMgmt], ApplianceTransportException
    + FullyQualifiedErrorId : HostnameAndCertDoNotMatch,Connect-HPOVMgmt
 

# 2. Put your code and/or captured output here.
# 3. Remove this comment block before submitting.
```

### Version Information
HPE OneView PowerShell Library Version (`Get-HPOVVersion` or `$PSLibraryVersion`): 
```powershell
LibraryVersion Path                                                                  
-------------- ----                                                                  
4.0.1554.2229  C:\Program Files\WindowsPowerShell\Modules\HPOneView.400\4.0.1554.2229
```
HPE OneView Appliance Version (`Get-HPOVVersion -ApplianceVer`):  
Output from `$PSVersionTable` on your Windows Host:

```PowerShell
# Output from $PSVersionTable

Name                           Value                                                                                                                                                                                                                 
----                           -----                                                                                                                                                                                                                 
PSVersion                      5.1.16299.98                                                                                                                                                                                                          
PSEdition                      Desktop                                                                                                                                                                                                               
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                                               
BuildVersion                   10.0.16299.98                                                                                                                                                                                                         
CLRVersion                     4.0.30319.42000                                                                                                                                                                                                       
WSManStackVersion              3.0                                                                                                                                                                                                                   
PSRemotingProtocolVersion      2.3                                                                                                                                                                                                                   
SerializationVersion           1.1.0.1

**Comment by ChrisLynchHPE** [comment]
I am unable to reproduce this condition.  I use an internal CA infrastructure for all of my lab virtual machines.  And to test, I just created a brand new self-signed certificate on my primary HPE OneView 4.0 appliance.  Can you please download openssl client, and then execute the following to verify the certificate on your appliance:

```
openssl s_client -showcerts -connect www.domain.com:443 
```
You can then look at the SAN values within the certificate using this script:
```PowerShell
$Appliance = 'MyAppliance.domain.com'

try 
{

    [System.Net.HttpWebRequest]$WebRequest = [System.Net.HttpWebRequest]::Create("https://$Appliance")

    $WebRequest.ServerCertificateValidationCallback = { $True }
    
    $Response = $WebRequest.GetResponse()

}

catch [Net.WebException] 
{ 

    if (-not($WebRequest.Connection) -and ([int]$Response.StatusCode -eq 0)) 
    {

        Write-Error $_.Exception.Message -Category ObjectNotFound -ErrorAction Stop

    } 

}

finally
{

    $Response.Dispose()

}

if ($null -ne $WebRequest.ServicePoint.Certificate) 
{

    # Get certificate object
    $Cert = [Security.Cryptography.X509Certificates.X509Certificate2]$WebRequest.ServicePoint.Certificate

    # Display SAN value from certificate
    $sanExt = $Cert.Extensions | Where-Object { $_.Oid.FriendlyName -match "subject alternative name" }
    $sanObjs = new-object -ComObject X509Enrollment.CX509ExtensionAlternativeNames
    $altNamesStr = [System.Convert]::ToBase64String($sanExt.RawData)
    $sanObjs.InitializeDecode(1, $altNamesStr)
    Foreach ($SAN in $sanObjs.AlternativeNames) { $SAN.strValue }

}

else
{

    Write-Host "Unable to connect to target server "

}
```

**Comment by ChrisLynchHPE** [comment]
Also, I have a verbose helper method that you can enable within the class library:
```PowerShell
[HPOneView.PKI.SslValidation]::ShowConsoleVerboseOutput = $true
```
Provide that after you load the module and before you issue the `Connect-HPOVMgmt` call.  There will be verbose output tagged with `[HPOneView.PKI.SslValidation]` that could also help.

**Comment by danlemire** [comment]
After performing your above provided steps, I don't see a mismatch between the hostname and the certificate. Before posting, we also regenerated the self-signed cert twice since the upgrade to 4.0. We started POC work with Greg Park back in October on v 3.0, and we performed the 4.0 update just a few days ago. Everything was working with 3.0 prior to the upgrade of Composer and the commandlets to v4.

I can provide actual values here if it matters to you, but for the record, I've only replace the hostname or fqdn with "host.domain.suffix"

.....

I connected via Open SSL as you suggested (using Bash on Windows, saved me from having to go install OpenSSL) to validate I have a valid certificate. 

openssl s_client -showcerts -connect host.domain.sufffix:443
CONNECTED(00000003)
depth=0 C = US, ST = California, L = Palo Alto, O = Hewlett Packard Enterprise, CN = hostname.domain.suffix
verify error:num=18:self signed certificate
verify return:1
depth=0 C = US, ST = California, L = Palo Alto, O = Hewlett Packard Enterprise, CN = hostname.domain.suffix
verify return:1
---
Certificate chain
 0 s:/C=US/ST=California/L=Palo Alto/O=Hewlett Packard Enterprise/CN=HOSTNAME.domain.suffix
   i:/C=US/ST=California/L=Palo Alto/O=Hewlett Packard Enterprise/CN=HOSTNAME.domain.suffix
-----BEGIN CERTIFICATE-----
Server certificate
subject=/C=US/ST=California/L=Palo Alto/O=Hewlett Packard Enterprise/CN=HOSTNAME.domain.suffix
issuer=/C=US/ST=California/L=Palo Alto/O=Hewlett Packard Enterprise/CN=HOSTNAME.domain.suffix
---
No client certificate CA names sent
Peer signing digest: SHA512
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 1730 bytes and written 487 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-SHA384
    Session-ID: EBCC0F4FC2CEC2912BE11C38FC9B2F9B21F6C6AFC230F3B1B52EA6C3716F6BB9
    Session-ID-ctx:
    Master-Key: B0CA25183CA950DAB1E6941A139EA59A07C7EE4EF30C92F515BC1A679096C371441C94FD3B1551245F6BAD655223509B
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 300 (seconds)
    TLS session ticket:
    0000 - 84 59 91 f2 ee 1c a7 1d-50 b7 bd 87 74 b6 9c 96   .Y......P...t...
    0010 - 2c c4 72 4f 0a 91 e8 62-c5 06 d2 ac 2c f1 31 30   ,.rO...b....,.10
    0020 - 5e 36 25 47 20 90 a0 d9-fc 7b d9 62 16 1f 2f 76   ^6%G ....{.b../v
    0030 - cf c8 2e f3 7a 1d cd ea-78 fd b7 1d 01 15 8a 02   ....z...x.......
    0040 - 45 cc 43 61 6f c1 ab 9e-ae 88 d0 a2 ce 47 69 b1   E.Cao........Gi.
    0050 - 77 52 ea d9 15 c0 0a a9-78 db 6e 84 22 9c 51 a9   wR......x.n.".Q.
    0060 - 8a de d2 f9 5b 33 c9 c8-e4 c2 16 f4 18 12 73 2b   ....[3........s+
    0070 - 01 4c f5 ad ad 93 ce 2d-e8 0e 23 d3 f6 87 87 13   .L.....-..#.....
    0080 - 90 41 bc e4 cb 96 07 1d-c7 e2 e0 76 d2 6a 19 2a   .A.........v.j.*
    0090 - 24 2f c2 73 89 b0 a4 06-0d e3 ec f1 42 4b 37 91   $/.s........BK7.
    00a0 - 1a 31 e7 0c 90 ac e2 90-25 5b 05 8c 38 0c d4 86   .1......%[..8...
    00b0 - 18 cb df 2e 90 11 91 fb-9a ab 36 80 4e 7c 58 66   ..........6.N|Xf

    Start Time: 1515794578
    Timeout   : 300 (sec)
    Verify return code: 18 (self signed certificate)


When running the PS command you provided for checking the SAN values, it shows the three values I expected.... 

HOSTNAME.domain.suffix
HOSTANAME
IP.Add.Re.ss

The helper method execution provides this additional information: 

VERBOSE: [HPOneView.PKI.SslValidation] Stating callback verification.
VERBOSE: [HPOneView.PKI.SslValidation] Cert has chain errors.
VERBOSE: [HPOneView.PKI.SslValidation] The hostname used to connect does not match the Subject or SAN of the provided host certificate.  Throw exception.
VERBOSE: [HPOneView.PKI.SslValidation] Stating callback verification.
VERBOSE: [HPOneView.PKI.SslValidation] Cert has chain errors.
VERBOSE: [HPOneView.PKI.SslValidation] The hostname used to connect does not match the Subject or SAN of the provided host certificate.  Throw exception.

**Comment by ChrisLynchHPE** [comment]
Hmm...  That certainly is how my appliance is setup.  Can you do the following:
```PowerShell
# Import module first
Import-Module HPOneView.400

# Enable class library verbose output
[HPOneView.PKI.SslValidation]::ShowConsoleDebugOutput = $true

# Attempt to connect to appliance
Connect-HPOVMgmt -Computername myappliance.domain.com -Credential $MyPSCredential -verbose

# Disable class library verbose output
[HPOneView.PKI.SslValidation]::ShowConsoleDebugOutput = $false
```
Enabling Debug here will show me more data about the method I use to parse the SAN value from the certificate.

**Comment by ChrisLynchHPE** [comment]
Are there any non-alphanumeric characters in the FQDN or hostname that you're trying to use?

**Comment by danlemire** [comment]
Ok, I did find and replace on HOSTNAME and HOSTNAME.domain.suffix. 

Everything looks good here, but the DNS names don't work. (I can use the IP, but names are important for me.

I appreciate your help Chris, here is output.

# Import module first
Import-Module HPOneView.400

# Enable class library verbose output
[HPOneView.PKI.SslValidation]::ShowConsoleDebugOutput = $true

# Attempt to connect to appliance
Connect-HPOVMgmt -Computername HOSTNAME.domain.suffix -Credential $MyPSCredential -verbose

# Disable class library verbose output
[HPOneView.PKI.SslValidation]::ShowConsoleDebugOutput = $false
VERBOSE: [CONNECT-HPOVMGMT] Bound PS Parameters: 
Name                           Value                                                                                                                                                                                                                 
----                           -----                                                                                                                                                                                                                 
Credential                     System.Management.Automation.PSCredential                                                                                                                                                                             
Verbose                        True                                                                                                                                                                                                                  
Hostname                       HOSTNAME.domain.suffix                                                                                                                                                                                          



VERBOSE: [CONNECT-HPOVMGMT] Creating Session Container
VERBOSE: [CONNECT-HPOVMGMT] Checking X-API Version
VERBOSE: [SEND-HPOVREQUEST] BEGIN
VERBOSE: [SEND-HPOVREQUEST] Bound PS Parameters: 
Key      Value                       
---      -----                       
Hostname HOSTNAME.domain.suffix
uri      /rest/version               



VERBOSE: [SEND-HPOVREQUEST] Process
VERBOSE: [SEND-HPOVREQUEST] Hostname value: HOSTNAME.domain.suffix

VERBOSE: [SEND-HPOVREQUEST] Filtering for Connection Object via String: HOSTNAME.domain.suffix
VERBOSE: [SEND-HPOVREQUEST] Processing 'HOSTNAME.domain.suffix' appliance connection request. 1 of 1
VERBOSE: [SEND-HPOVREQUEST] Requested URI '/rest/version' to 'HOSTNAME.domain.suffix'
VERBOSE: [SEND-HPOVREQUEST] We have reached the URI Whitelist condition block. Unauth request allowed for '/rest/version'.
VERBOSE: [RESTCLIENT] Building new [System.Net.HttpWebRequest] object for GET https://HOSTNAME.domain.suffix/rest/version
VERBOSE: [SEND-HPOVREQUEST] Request: GET https://HOSTNAME.domain.suffix/rest/version
VERBOSE: [SEND-HPOVREQUEST] Request Header 1: User-Agent = HPEOneview_PowerShell/4.00
VERBOSE: [SEND-HPOVREQUEST] Request Header 2: Content-Type = application/json
VERBOSE: [SEND-HPOVREQUEST] Request Header 3: Accept = application/json
VERBOSE: [SEND-HPOVREQUEST] Request Header 4: X-API-Version = 600
VERBOSE: [SEND-HPOVREQUEST] Request Header 5: accept-language = en_US
VERBOSE: [SEND-HPOVREQUEST] Request Header 6: accept-encoding = gzip, deflate
[DEBUG]: [HPOneView.PKI.SslValidation] Init parse SAN from certificate.
[DEBUG]: [HPOneView.PKI.SslValidation] Certificate contains Subject Alternative Names. Parsing.
[DEBUG]: [HPOneView.PKI.SslValidation] Adding DNS Name=HOSTNAME.domain.suffix to collection
[DEBUG]: [HPOneView.PKI.SslValidation] Adding DNS Name=HOSTNAME to collection
[DEBUG]: [HPOneView.PKI.SslValidation] Adding IP Address=28.157.235.60 to collection
[DEBUG]: [HPOneView.PKI.SslValidation] Adding DNS Name=28.157.235.60 to collection
[DEBUG]: [HPOneView.PKI.SslValidation] Init parse SAN from certificate.
[DEBUG]: [HPOneView.PKI.SslValidation] Certificate contains Subject Alternative Names. Parsing.
[DEBUG]: [HPOneView.PKI.SslValidation] Adding DNS Name=HOSTNAME.domain.suffix to collection
[DEBUG]: [HPOneView.PKI.SslValidation] Adding DNS Name=HOSTNAME to collection
[DEBUG]: [HPOneView.PKI.SslValidation] Adding IP Address=28.157.235.60 to collection
[DEBUG]: [HPOneView.PKI.SslValidation] Adding DNS Name=28.157.235.60 to collection
VERBOSE: [SEND-HPOVREQUEST] Net.WebException Error caught
VERBOSE: [SEND-HPOVREQUEST] Exception Object: 

PSMessageDetails      : 
Exception             : System.Net.WebException: The underlying connection was closed: An unexpected error occurred on a receive. ---> HPOneView.Library.ApplianceTransportException: The hostname used to connect does not match the Subject or SAN 
                        of the provided host certificate.  Throw exception.
                           at HPOneView.PKI.SslValidation.CertificateValidationCallBack(Object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
                           at System.Net.ServerCertValidationCallback.Callback(Object state)
                           at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
                           at System.Net.ServerCertValidationCallback.Invoke(Object request, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
                           at System.Net.Security.SecureChannel.VerifyRemoteCertificate(RemoteCertValidationCallback remoteCertValidationCallback, ProtocolToken& alertToken)
                           at System.Net.Security.SslState.CompleteHandshake(ProtocolToken& alertToken)
                           at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)
                           at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
                           at System.Net.TlsStream.ProcessAuthentication(LazyAsyncResult result)
                           at System.Net.TlsStream.Write(Byte[] buffer, Int32 offset, Int32 size)
                           at System.Net.ConnectStream.WriteHeaders(Boolean async)
                           --- End of inner exception stack trace ---
                           at System.Net.HttpWebRequest.GetResponse()
                           at CallSite.Target(Closure , CallSite , WebRequest )
TargetObject          : 
CategoryInfo          : NotSpecified: (:) [], WebException
FullyQualifiedErrorId : WebException
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Send-HPOVRequest<Process>, C:\Program Files\WindowsPowerShell\Modules\HPOneView.400\4.0.1554.2229\HPOneView.400.psm1: line 4702
                        at Connect-HPOVMgmt<Begin>, C:\Program Files\WindowsPowerShell\Modules\HPOneView.400\4.0.1554.2229\HPOneView.400.psm1: line 6459
                        at <ScriptBlock>, <No file>: line 8
PipelineIterationInfo : {}




VERBOSE: [SEND-HPOVREQUEST] Exception Message: The underlying connection was closed: An unexpected error occurred on a receive.
VERBOSE: [SEND-HPOVREQUEST] InnerException FullyQualifiedErrorId: ApplianceTransportException
VERBOSE: [SEND-HPOVREQUEST] InnerException Message: The hostname used to connect does not match the Subject or SAN of the provided host certificate.  Throw exception.
VERBOSE: [NEW-ERRORRECORD] Building ErrorRecord object
VERBOSE: [SEND-HPOVREQUEST] Cleaning up HttpWebRequest
Connect-HPOVMgmt : The hostname used to connect does not match the Subject or SAN of the provided host certificate.  Throw exception.
At line:8 char:1
+ Connect-HPOVMgmt -Computername HOSTNAME.domain.suffix -Credenti ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (Hostname:String) [Connect-HPOVMgmt], ApplianceTransportException
    + FullyQualifiedErrorId : HostnameAndCertDoNotMatch,Connect-HPOVMgmt

**Comment by ChrisLynchHPE** [comment]
As I can't seem to reproduce this in other environments, I am marking this as a bug, with Low Severity.  I have done some code refactoring in the Class Library I ship with the module.  I will have this included with the next update.

**Comment by ChrisLynchHPE** [comment]
This was addressed in the [Release 4.00.1587.3400 ](https://github.com/HewlettPackard/POSH-HPOneView/releases/tag/v4.00.1587.3400 ).

Closing.
